tests/unit/test_retry.py

```diff
@@ -467,3 +490,425 @@ def test___init___when_retry_is_executed(self, sleep, uniform):
         assert _some_function.call_count == 2
         target.assert_has_calls([mock.call("meep"), mock.call("meep")])
         sleep.assert_any_call(retry_._initial)
+
+    def _generator_mock(
+        self,
+        num=5,
+        error_on=None,
+        return_val=None,
+        exceptions_seen=None,
+        ignore_sent=False,
+    ):
+        """
+        Helper to create a mock generator that yields a number of values
+        Generator can optionally raise an exception on a specific iteration
+        """
+        try:
+            sent_in = None
+            for i in range(num):
+                if error_on and i == error_on:
+                    raise ValueError("generator mock error")
+                sent_in = yield (sent_in if sent_in else i)
+                if ignore_sent:
+                    sent_in = None
+            return return_val
+        except (Exception, BaseException, GeneratorExit) as e:
+            # keep track of exceptions seen by generator
+            if exceptions_seen is not None:
+                exceptions_seen.append(e)
+            raise
+
+    @mock.patch("time.sleep", autospec=True)
+    def test___call___generator_success(self, sleep):
+        """
+        Test that a retry-decorated generator yields values as expected
+        This test checks a generator with no issues
+        """
+        import types
+        import collections
+
+        retry_ = retry.Retry()
+
+        decorated = retry_(self._generator_mock)
+
+        num = 10
+        result = decorated(num)
+        # check types
+        assert isinstance(decorated(num), collections.abc.Iterable)
+        assert isinstance(decorated(num), types.GeneratorType)
+        assert isinstance(self._generator_mock(num), collections.abc.Iterable)
+        assert isinstance(self._generator_mock(num), types.GeneratorType)
+        # check yield contents
+        unpacked = [i for i in result]
+        assert len(unpacked) == num
+        for a, b in zip(decorated(num), self._generator_mock(num)):
+            assert a == b
+        sleep.assert_not_called()
+
+    @mock.patch("time.sleep", autospec=True)
+    def test___call___generator_retry(self, sleep):
+        """
+        Tests that a retry-decorated generator will retry on errors
+        """
+        on_error = mock.Mock(return_value=None)
+        retry_ = retry.Retry(
+            on_error=on_error,
+            predicate=retry.if_exception_type(ValueError),
+            is_stream=True,
+            timeout=None,
+        )
+        result = retry_(self._generator_mock)(error_on=3)
+        # error thrown on 3
+        # generator should contain 0, 1, 2 looping
+        unpacked = [next(result) for i in range(10)]
+        assert unpacked == [0, 1, 2, 0, 1, 2, 0, 1, 2, 0]
```

Comment:
Where does `_generator_mock.i` get reset when you `raise ValueError(…)`?

Reply:
This is testing the core functionality of the streaming retries implementation. When the target generator raises an exception, a new instance is [swapped out under the hood](https://github.com/googleapis/python-api-core/blob/8cc6ea9c4f9717a2a0b0c8e2bd55efc726f5837e/google/api_core/retry_streaming.py#L104), without disrupting the wrapper stream